### PR TITLE
[LTC] add shape info to types in TS lowering

### PR DESCRIPTION
**Pull request by makslevental**
effect of this changes looks like this

```
graph(%p0 : Float(10, strides=[1]),
      %p1 : Float(10, 500, strides=[500, 1]),
      %p2 : Float(500, strides=[1]),
      %p3 : Float(500, 800, strides=[800, 1]),
      %p4 : Float(1, 50, 4, 4, strides=[800, 16, 4, 1])):
  %2 : int[] = prim::Constant[value=[1, 0]]()
  %3 : int[] = prim::Constant[value=[1, 0]]()
  %4 : Float(500, 10, strides=[10, 1]) = aten::permute(%p1, %3)
  %7 : int[] = prim::Constant[value=[1, 0]]()
  %8 : int[] = prim::Constant[value=[1, 0]]()
  %9 : Float(800, 500, strides=[500, 1]) = aten::permute(%p3, %8)
  %11 : int[] = prim::Constant[value=[1, 800]]()
  %12 : int[] = prim::Constant[value=[1, 800]]()
  %13 : Float(1, 800, strides=[800, 1]) = aten::reshape(%p4, %12)
  %14 : int = prim::Constant[value=1]()
  %15 : int = prim::Constant[value=1]()
  %16 : Float(1, 500, strides=[0, 0]) = aten::addmm(%p2, %13, %9, %14, %15)
  %17 : Float(1, 500, strides=[0, 0]) = aten::relu(%16)
  %18 : int = prim::Constant[value=1]()
  %19 : int = prim::Constant[value=1]()
  %20 : Float(1, 10, strides=[0, 0]) = aten::addmm(%p0, %17, %4, %18, %19)
  return (%20)
```

I will add tests if before landing if this is the right approach.

Note that I did find [ForEachMutableSubshape](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensors/shape_util.h#L149) but as far as I understand this I'm not changing the shapes per se but the types.

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 1]
```diff
@@ -58,8 +60,33 @@ class TSNodeLowering : public NodeLowering {
       return false;
     }
     LTC_CHECK_EQ(node->num_outputs(), ops.size());
-    for (size_t i = 0; i < ops.size(); ++i) {
-      loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+    if (lazy_tensors::sys_util::GetEnvBool("LTC_IR_TS_SHAPE_TYPES", false)) {
```
Comment (makslevental) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:63]: not sure if putting behind env is good or bad

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 2]
```diff
@@ -58,8 +60,33 @@ class TSNodeLowering : public NodeLowering {
...
+      for (size_t i = 0; i < ops.size(); ++i) {
+        loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+      }
+    } else {
+      auto type_transformer = [](const Shape& shape, const at::TypePtr& type) {
+        return type->expect<at::TensorType>()
+            ->withScalarType(TensorTypeFromLtcType(shape.element_type()))
+            ->withSizesStrides(shape.as_sizes(), ComputeShapeStrides(shape));
```
Comment (makslevental) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:71]: not sure if this covers everything (though it does like it does from the TS IR generated)

Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:71]: hhm, I'm not seeing here
* `device`
* `requires_grad` (which we may not care about)

You could use one of `TensorType::create` constructors to avoid constructing a type, one property at a time.

` ComputeShapeStrides(shape)` do we know if this handles transposed strides correctly?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 1 above)
Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:63]: It's probably okay for now, although a python switch like this one: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/init.cpp#L608

would seem cleaner and more flexible.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 3]
```diff
@@ -58,8 +60,33 @@ class TSNodeLowering : public NodeLowering {
...
+      };
+      for (size_t i = 0; i < ops.size(); ++i) {
+        auto shape = node->shape(i);
+        if (shape.IsArray()) {
+          ops[i]->setType(type_transformer(shape, ops[i]->type()));
+        } else if (shape.IsTuple()) {
```
Comment (Krovatkin) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:77]: I'm trying to think of aten ops that would return nested tuples or nested arrays or any combination of the two. I guess, until then we  can rely on that assert in `expect<TensorType>`  in `type_transformer`

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 1 above)
Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:63]: nitpick: do you need to prefix `sys_util` if you already included `using namespace lazy_tensors` ?

**Review by Krovatkin — COMMENTED** [review]
a few questions.
would be nice to see more tests.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 3 above)
Comment (makslevental) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:77]: I think @asuhan said that there are no nested tuples and I think that "Array" here means tensor: actually https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensors/primitive_util.cc#L6

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 1 above)
Comment (makslevental) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:63]: re the python switch: I think that should go in `LoweringContext` but i'm not sure since there's no such flags being passedyet.
re `sys_util` yes the namespace resolution doesn't but actually I didn't need the `using`.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 2 above)
Comment (makslevental) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:71]: my thought in using the `with*` properties was to preserve everything about the original type except for what I can get from the shape. so e.g. `device` and `requires_grad` should be preserved right? 

https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/aten/src/ATen/core/jit_type.h#L569

The fact that those attributes aren't rendered in the type is due to how the JIT serializes the graph I think? 

https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp#L70

re `ComputeShapeStrides` i'm not sure but maybe `minor_to_major` here implies yes?

https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/tensor_util.cpp#L704

**Comment by makslevental** [comment]
one thing i was trying to test is the tuple handling but for some reason i could not make it hit that code path no matter what i put into my model, so .e.g 

```
    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2, 2)
        x = x.view(-1, 4*4*50)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        x = torch.max(x, dim=1)
        return x
```

does not trigger that path. I'm not sure if I've misunderstood `PrimitiveType::Tuple` or `torch.max` ...

**Comment by asuhan** [comment]
@makslevental We're currently using interop for max_pool* operations, so they break the trace. Try a batch norm instead, that should give you tuples, but the C++ tests cover it anyway.

For reference, this is what we do for max pool: https://github.com/pytorch/pytorch/blob/ee1c705c1c16a47842638db027f8470397edcac7/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_autograd_ops_ts.cpp#L12. The max pool family of operators is slightly weird in that we don't just use the auto-generated interop and we have to do this instead.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 3 above)
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:75]: Correct, "Array" is terminology inherited from TF. We should cleanup these quirky terms at some point. I can confirm there won't be nested tuples, there's no operator which would require them to represent its types.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 2 above)
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:69]: FWIW I don't think we'd have a use for either in the generated TS graphs, we use TS-as-a-backend here. The model of operation is that the device stays the same as the inputs device and we don't care about gradients either at this point, it's just a computation backend at this point and differentiation has been taken care of by autograd, TL;DR I wouldn't worry about preserving anything but shapes.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 4]
```diff
@@ -58,8 +58,33 @@ class TSNodeLowering : public NodeLowering {
       return false;
     }
     LTC_CHECK_EQ(node->num_outputs(), ops.size());
-    for (size_t i = 0; i < ops.size(); ++i) {
-      loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+    if (!lazy_tensors::sys_util::GetEnvBool("LTC_IR_TS_SHAPE_TYPES", false)) {
```
Comment (asuhan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:61]: You can just use `if (!Shape::IsDynamicMode())` here instead. If we're in static shape mode, no one should mind having the fully refined types in the generated TS anyway.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 5]
```diff
@@ -58,8 +58,33 @@ class TSNodeLowering : public NodeLowering {
...
+      for (size_t i = 0; i < ops.size(); ++i) {
+        loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+      }
+    } else {
+      auto type_transformer = [](const lazy_tensors::Shape& shape, const at::TypePtr& type) {
+        return type->expect<at::TensorType>()
+            ->withScalarType(TensorTypeFromLtcType(shape.element_type()))
+            ->withSizesStrides(shape.as_sizes(), ComputeShapeStrides(shape));
+      };
+      for (size_t i = 0; i < ops.size(); ++i) {
```
Comment (asuhan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:71]: Maybe move this loop to a helper? Up to you since the function isn't that big.

**Review by asuhan — CHANGES_REQUESTED** [review]
Generally looks good, added comments inline.

Could you check that the tuple path is hit by any of `AtenLtcTsTensorTest.TestBatchNorm*` tests or any `AtenLtcTsTensorTest.*` in general? I'm pretty sure it is, I would double check we don't lack test coverage for this.

**Comment by makslevental** [comment]
@Krovatkin I'm gonna punt on the tests for this one. I started writing them and had to expose a bunch of things of things that aren't public or `TORCH_API` (like `RefineTypesInLoweredOp` and `LowerToTS` and `GetTSNodeLowering`) and I don't think it's worth breaking encapsulation just to test such a simple function :p

**Comment by wconstab** [comment]
I'd like to pull @Krovatkin into this conversation and make sure that the design intent is aligned with what we're doing to enable dynamic shape support.

**Comment by makslevental** [comment]
>I am not sure what the context of the original effort around this PR was

The original intent was as it says on the tin: to have a way of recovering shape from LTC. The context was memory planning, which requires those shapes for inferring buffer/allocation sizes (you might recall my intern project last summer @wconstab). That project was scoped to originally go through LTC but, partially since this never landed, we ended up just using TS IR. 

I can't speak to the current priorities re memory planning (I'm vaguely aware it's being worked in, in parallel, within NNC) and therefore can't speak to the value of this PR. But if it's useful, I'm happy to get it finish it up and incorporate any further changes (all of asuhan's changes were incorporated at the time).
